todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/domain/logic.kt

```diff
@@ -0,0 +1,140 @@
+ * limitations under the License.
+ * -/-/-
+ */
+
+package com.example.android.architecture.blueprints.todoapp.tasks.domain
+
+import com.example.android.architecture.blueprints.todoapp.tasks.domain.FeedbackType.*
+import com.example.android.architecture.blueprints.todoapp.util.change
+import com.example.android.architecture.blueprints.todoapp.util.without
+import com.spotify.mobius.Effects.effects
+import com.spotify.mobius.First
+import com.spotify.mobius.First.first
+import com.spotify.mobius.Next
+import com.spotify.mobius.Next.*
+
+
+fun init(model: TasksListModel) : First<TasksListModel, TasksListEffect> =
+        model.tasks?.let { first(model, effects(LoadTasks as TasksListEffect)) }
+                ?: first(model.copy(loading = true), effects(RefreshTasks, LoadTasks))
+
+fun update(model: TasksListModel, event: TasksListEvent): Next<TasksListModel, TasksListEffect> =
+        when (event) {
+            RefreshRequested -> onRefreshRequested(model)
+            NewTaskClicked -> onNewTaskClicked()
+            is NavigateToTaskDetailsRequested -> onNavigateToTaskDetailsRequested(model, event)
+            is TaskMarkedComplete -> onTaskCompleted(model, event)
+            is TaskMarkedActive -> onTaskActivated(model, event)
+            ClearCompletedTasksRequested -> onCompletedTasksCleared(model)
+            is FilterSelected -> onFilterSelected(model, event)
+            is TasksLoaded -> onTasksLoaded(model, event)
+            TaskCreated -> onTaskCreated()
+            TasksRefreshed -> onTasksRefreshed(model)
+            TasksFreshFailed -> onTasksRefreshFailed(model)
+            TasksLoadingFailed -> onTasksLoadingFailed(model)
+        }
+
+private fun onRefreshRequested(model: TasksListModel): Next<TasksListModel, TasksListEffect> =
+        next(model.copy(loading = true), effects(RefreshTasks))
+
+private fun onNewTaskClicked(): Next<TasksListModel, TasksListEffect> =
+        dispatch(effects(StartTaskCreationFlow))
+
+private fun onNavigateToTaskDetailsRequested(model: TasksListModel,
+                                             event: NavigateToTaskDetailsRequested)
+        : Next<TasksListModel, TasksListEffect> =
+        model.findTaskById(event.taskId)?.let {
+            dispatch<TasksListModel, TasksListEffect>(effects(NavigateToTaskDetails(it)))
+        }
+        ?: throw IllegalStateException("Task does not exist")
+
+private fun onTaskCompleted(model: TasksListModel,
+                            event: TaskMarkedComplete): Next<TasksListModel, TasksListEffect> =
+        model.findTaskIndexById(event.taskId)
+                .run {
+                    updateTask(copy(second = second.complete()), model, MARKED_COMPLETE)
+                }
+
+private fun onTaskActivated(model: TasksListModel,
+                            event: TaskMarkedActive): Next<TasksListModel, TasksListEffect> =
+        model.findTaskIndexById(event.taskId)
+                .run {
```

Comment:
The use of run here does not really make the function shorter or more readable, I would prefer it without

Reply:
I was kinda going for making everything an expression and avoiding variables :)